packages/smart-contracts/src/contracts/ERC20SwapToConversion.sol

```diff
@@ -1,160 +1,232 @@
-  * @param _erc20Address Address of an ERC20 used for payment
-  */
-  function approveRouterToSpend(address _erc20Address) public {
-    IERC20 erc20 = IERC20(_erc20Address);
-    uint256 max = 2**256 - 1;
-    erc20.safeApprove(address(swapRouter), max);
-  }
-
-  /*
-  * Admin functions to edit the admin, router address or proxy address
-  */
-  function setPaymentProxy(address _paymentProxyAddress) public onlyOwner {
-    paymentProxy = Erc20ConversionProxy(_paymentProxyAddress);
-    chainlinkConversionPath = ChainlinkConversionPath(paymentProxy.chainlinkConversionPath());
-  }
+    /**
+     * @notice Authorizes the swap router to spend a new payment currency (ERC20).
+     * @param _erc20Address Address of an ERC20 used for payment
+     */
+    function approveRouterToSpend(address _erc20Address) public {
+        IERC20 erc20 = IERC20(_erc20Address);
+        uint256 max = 2**256 - 1;
+        erc20.safeApprove(address(swapRouter), max);
+    }
 
-  function setRouter(address _newSwapRouterAddress) public onlyOwner {
-    swapRouter = IUniswapV2Router02(_newSwapRouterAddress);
-  }
+    /*
+     * Admin functions to edit the router address, the fees amount and the fees collector address
+     */
+    function setRouter(address _newSwapRouterAddress) public onlyOwner {
+        swapRouter = IUniswapV2Router02(_newSwapRouterAddress);
+    }
 
-  /*
-  * Internal functions to reduce the stack in swapTransferWithReference()
-  */
-  function getConversion(
-    address[] memory _path,
-    uint256 _requestAmount,
-    uint256 _requestFeeAmount
-  )
-    internal
-    view
-    returns (uint256 conversion)
-  {
-    (conversion, ) = chainlinkConversionPath.getConversion(_requestAmount + _requestFeeAmount, _path);
-  }
+    function setRequestSwapFees(uint256 _newRequestSwapFees) public onlyOwner {
+        requestSwapFees = _newRequestSwapFees;
+    }
+
+    function setRequestFeesCollector(address _newRequestFeesCollector) public onlyOwner {
+        requestFeesCollector = _newRequestFeesCollector;
+    }
+
+    /*
+     * Internal functions to reduce the stack in swapTransferWithReference()
+     */
+    function _getConversion(
+        address[] memory _path,
+        uint256 _requestAmount,
+        uint256 _requestFeeAmount
+    ) internal view returns (uint256 conversion) {
+        (conversion, ) = chainlinkConversionPath.getConversion(
+            _requestAmount + _requestFeeAmount,
+            _path
+        );
+    }
+
+    /**
+     * Internal functions to reduce the stack in swapTransferWithReference()
+     * Approve the SwapRouter to spend the spentToken if needed
+     * Swap the spentToken in exchange for the expectedToken
+     * Approve the payment proxy to spend the expected token if needed.
+     */
+    function _swapAndApproveIfNeeded(
+        address _paymentProxy,
+        uint256 _amountInMax, // SpentToken
+        address[] memory _uniswapPath, // from spentToken to expectedToken on uniswap
```

Comment:
```suggestion
        address[] memory _swapPath, // from spentToken to expectedToken on the swap router contract
```
Please clean all references of this kind as Uniswap is not present on every network but we can find equivalents with the same interface and good liquidity.